finance_dl/amazon.py

```diff
@@ -273,11 +277,21 @@ def retrieve_invoices(self, invoice_hrefs):
             # Wait until it is all generated.
             def get_source():
                 source = self.driver.page_source
-                if 'Grand Total:' in source:
+                if 'Grand Total:' in source or 'Order Canceled' in source:
                     return source
+                elif 'problem loading this order' in source:
+                    return True
+                elif self.find_visible_elements(By.XPATH, '//input[@type="password"]'):
+                    self.finish_login()
+                    return get_source()
                 return None
```

Comment:
This doesn't look right.
This method should either return the source, or True indicating failure. 
It seems that if the page is loading and the first three clauses are all False, then None is returned and not handled.
I also suggest using None to indicate failure, rather than True.

Reply:
This is a `wait_and_return` callback; `wait_and_return` interprets a None return as a signal to keep waiting and try again. I guess the more pythonic thing to do would be to raise, though; done.

I've also swapped the recursion for a None return, since using two different techniques to signal a retry is confusing at best.